File: ros2_ws/src/cpp_pubsub/scripts/interp_test.py
from math import pi, cos
import matplotlib.pyplot as plt


def cosineInterpolate(start, end, npoints):
    new_points = []
    for index in range(npoints):
        # angle = (start + i*interval) * pi
        angle = index / (npoints-1) * pi
        new_ratio = (1.0 - cos(angle)) * 0.5   # in [0, 1]
        new_point = start + new_ratio * (end - start)   # in [start, end]
        new_points.append(new_point)
    return new_points


res_list = cosineInterpolate(1, 5, 10)

print(res_list)

plt.plot(res_list)
plt.show()

File: ros2_ws/src/cpp_pubsub/backups/interp_test.cpp
#include <stdio.h>
#include <stdlib.h>
#include <vector>
#include <iostream>
#include <math.h>
// #include <correct_math_defines.h>


// Function to cosine interpolate a vector and return a new vector
std::vector<double> cosine_interpolate(double start, double end, int npoints) {
    std::vector<double> new_points = {};
    for (int index=0; index<npoints; index++) {
        double angle = (double)index / (double)(npoints-1) * 3.1415926;
        double new_ratio = (1.0 - cos(angle)) * 0.5;            // in [0, 1]
        double new_point = start + new_ratio * (end - start);   // in [start, end]
        new_points.push_back(new_point);
    }
    return new_points;
}


void print_vector(std::vector<double>& vec) {
  
  std::cout << "[ ";
  for (unsigned int i=0; i<vec.size(); i++) {
    std::cout << vec.at(i) << ' ';
  }
  std::cout << "]" << std::endl;
}


int main(int argc, char *argv[])
{
    std::vector<double> points = cosine_interpolate(1, 5, 10);
    print_vector(points);

    return 0;
}